src/System.Security.Cryptography.Cng/src/Internal/Cryptography/CngAsymmetricAlgorithmCore.cs

```diff
@@ -0,0 +1,144 @@
+            {
+                _lazyKey.Dispose();
+                _lazyKey = null;
+            }
+
+            // If we don't have a key yet, we need to generate a random one now.
+            if (_lazyKey == null)
+            {
+                CngKeyCreationParameters creationParameters = new CngKeyCreationParameters()
+                {
+                    ExportPolicy = CngExportPolicies.AllowPlaintextExport,
+                };
+
+                CngProperty keySizeProperty = new CngProperty(KeyPropertyName.Length, BitConverter.GetBytes(keySize), CngPropertyOptions.None);
+                creationParameters.Parameters.Add(keySizeProperty);
+                _lazyKey = CngKey.Create(algorithm, null, creationParameters);
+            }
+
+            return _lazyKey;
+        }
+
+        public void SetKey(CngKey key)
+        {
+            Debug.Assert(key != null);
+
+            // If we already have a key, clear it out.
+            if (_lazyKey != null)
+            {
+                _lazyKey.Dispose();
+            }
+
+            _lazyKey = key;
+        }
+
+        public static byte[] HashData(byte[] data, int offset, int count, HashAlgorithmName hashAlgorithm)
+        {
+            // we're sealed and the base should have checked this already.
+            Debug.Assert(data != null);
+            Debug.Assert(offset >= 0 && offset <= data.Length);
+            Debug.Assert(count >= 0 && count <= data.Length);
+            Debug.Assert(!string.IsNullOrEmpty(hashAlgorithm.Name));
+
+            using (HashProviderCng hashProvider = new HashProviderCng(hashAlgorithm.Name, null))
+            {
+                hashProvider.AppendHashData(data, offset, count);
+                byte[] hash = hashProvider.FinalizeHashAndReset();
+                return hash;
+            }
+        }
+
+        public static byte[] HashData(Stream data, HashAlgorithmName hashAlgorithm)
+        {
+            // We're sealed and the base should have checked these already.
+            Debug.Assert(data != null);
+            Debug.Assert(!string.IsNullOrEmpty(hashAlgorithm.Name));
+
+            using (HashProviderCng hashProvider = new HashProviderCng(hashAlgorithm.Name, null))
+            {
+                // Default the buffer size to 4K.
+                byte[] buffer = new byte[4096];
+                int bytesRead;
+                while ((bytesRead = data.Read(buffer, 0, buffer.Length)) > 0)
+                {
+                    hashProvider.AppendHashData(buffer, 0, bytesRead);
+                }
+                byte[] hash = hashProvider.FinalizeHashAndReset();
+                return hash;
+            }
+        }
+
+        public static unsafe byte[] SignHash(CngKey key, byte[] hash, AsymmetricPaddingMode paddingMode, void* pPaddingInfo, int estimatedSize)
+        {
+#if DEBUG
+            estimatedSize = 2;  // Make sure the NTE_BUFFER_TOO_SMALL scenario gets exercised.
```

Comment:
There are only two build flavors so our ability to auto-coverage this way is limited.

As a practical matter, retail builds do the just-right case on every EcDsaCng.SignHash() and the over-large case on (almost) every RsaCng.Decrypt().